src/crimson/osd/pg_backend.cc

```diff
@@ -935,8 +935,10 @@ PGBackend::omap_get_keys(
     std::min(max_return, local_conf()->osd_max_omap_entries_per_request);
 
   // TODO: truly chunk the reading
-  return maybe_get_omap_vals(store, coll, os.oi, start_after).safe_then_interruptible(
-    [=, &osd_op](auto ret) {
+  return seastar::do_with(std::move(start_after),
```

Comment:
@liu-chunmei i don't quite follow on your change.  before your change,`start_after` is passed to `maybe_get_omap_vals()` by reference, which in turn calls into `FuturizedStore::omap_get_values()`. where a temporary instance of `optional<string>`. but in the implementation of seastore, we have

https://github.com/ceph/ceph/blob/596e1ca370dbf5c70a4d8bf5fdfdfdd9eeeead1b/src/crimson/os/seastore/seastore.cc#L344-L361

the variables are "captured" by reference, see https://github.com/scylladb/seastar/blob/fb183d2c661d8fb988d9cd46809be7214d06488f/include/seastar/core/do_with.hh#L126-L132, which calls into
 https://github.com/scylladb/seastar/blob/fb183d2c661d8fb988d9cd46809be7214d06488f/include/seastar/core/do_with.hh#L80-L85

please note, `std::forward_as_tuple()` constructs a tuple of references instead of a tuple of values. that's why we end up with `do_with()` a bunch of references in Seastore, even with `do_with()`.

when it comes to ailenstore, we have

https://github.com/ceph/ceph/blob/596e1ca370dbf5c70a4d8bf5fdfdfdd9eeeead1b/src/crimson/os/alienstore/alien_store.cc#L353-L375

the variables used by the continuation in `do_with()` are captured by **value**.

so i think a better fix would be

1. change the `FuturizedStore::omap_get_values()`, so it accepts a `std::optional<std::string>&& start` instead of `const std::optional<std::string> &start`, and
2. capture it by move in both of its implementations.

and we might need to keep in mind not to `do_with()` references in hope to extend their life cycle -- we will end up with dangling references this way.

Reply:
I think the Seastore's overload is this one:

```cpp
SeaStore::omap_get_values_ret_t SeaStore::omap_get_values(
  CollectionRef ch,
  const ghobject_t &oid,
  const std::optional<string> &start)
{
  return seastar::do_with(oid, start,
                          [this, ch=std::move(ch)](auto& oid, auto& start) {
    return omap_list(
      ch, oid, start,
      OMapManager::omap_list_config_t::with_inclusive(false));
  });
}
```

So, `do_with()` is passed with a const l-value reference to the `optional<string>`. I like they idea of taking `&&`.